### PR TITLE
eds: consume envoy::api::v2::core::HealthStatus settings for endpoints.

### LINKED ISSUE #2726: EDS delivered health status
We added the ability to deliver health status for endpoints in the v2 API (https://github.com/envoyproxy/data-plane-api/blob/818dc1db99f73830b4f9be50c243d439e23555d6/envoy/api/v2/endpoint/endpoint.proto#L31). This can be used to provide centralized health checking and influence gentle failover between priority groups.

We have a requirement for this but it's not yet implemented, so will take this on.

### PR DESCRIPTION
**Pull request by htuch**
Fixes #2726.

Risk Level: Low (no change unless the health status is explicitly set in
EDS response).
Testing: unit tests for eds.cc/upstream_impl.cc behavior, integration
test to validate end-to-end via eds_integration_test.cc.
Release Notes: https://github.com/envoyproxy/data-plane-api/pull/591

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/common/upstream/upstream_impl.cc [hunk 1]
```diff
@@ -533,6 +533,7 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
                                                    HostVector& hosts_added,
                                                    HostVector& hosts_removed, bool depend_on_hc) {
   uint64_t max_host_weight = 1;
+  bool eds_health_changed = false;
```
Comment (alyssawilk) [inline source/common/upstream/upstream_impl.cc]: If the host is already unhealthy, do we need to propogate that the host is now unhealthy for 2 reasons?
I'm assuming the answer is yes, but I'd like to understand why, and encourage you to comment in the code for future curious people.

File: test/integration/eds_integration_test.cc [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+TEST_P(EdsIntegrationTest, HealthUpdate) {
+  initialize();
+  // Initial state, no cluster members.
+  EXPECT_EQ(1, test_server_->counter("cluster.cluster_0.update_success")->value());
+  EXPECT_EQ(0, test_server_->counter("cluster.cluster_0.membership_change")->value());
+  EXPECT_EQ(0, test_server_->gauge("cluster.cluster_0.membership_total")->value());
+  EXPECT_EQ(0, test_server_->gauge("cluster.cluster_0.membership_healthy")->value());
+  // 2/2 healthy endpoints.
+  updateEndpoints(2, 2);
+  test_server_->waitForCounterGe("cluster.cluster_0.update_success", 2);
```
Comment (alyssawilk) [inline test/integration/eds_integration_test.cc]: We had a similar pattern with the load tests, and it became problematic as we added more tests as each update required manually increment each integer.  Can you change this either to something like 

waitForCounterGe("...update_success", update_successes++) 

or wrapper the whole thing

s/updateEndpoints/setEndpoints/ and add
void updateEndpoints(int endpoints, int healthy) {
  setEndpoints(endpoints, healthy);
 test_server_->waitForCounterGe(..... update_counter_++);
}

File: test/integration/eds_integration_test.cc [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+      // Switch predefined cluster_0 to EDS filesystem sourcing.
+      auto* cluster_0 = bootstrap.mutable_static_resources()->mutable_clusters(0);
+      cluster_0->mutable_hosts()->Clear();
+      cluster_0->set_type(envoy::api::v2::Cluster::EDS);
+      auto* eds_cluster_config = cluster_0->mutable_eds_cluster_config();
+      eds_cluster_config->mutable_eds_config()->set_path(eds_path_);
+    });
+    HttpIntegrationTest::initialize();
+    for (uint32_t i = 0; i < upstream_endpoints_; ++i) {
```
Comment (alyssawilk) [inline test/integration/eds_integration_test.cc]: What would you think of extending the test framework to have a configurable number of upstreams?  I think that'd be a net win.

File: source/common/upstream/upstream_impl.cc (hunk 1 above)
Comment (htuch) [reply source/common/upstream/upstream_impl.cc]: I can add this optimization, it's pretty straightforward.

Comment (alyssawilk) [reply source/common/upstream/upstream_impl.cc]: s/eds_health_changed/health_changed ?

Has the EDS health status changed of any endpoint? -> 
Has the EDS health status changed the health of any endpoint?

**Pull request merged**